ocaml/xapi/xapi_pgpu_helpers.ml

```diff
@@ -136,82 +137,108 @@ let assert_capacity_exists_for_VGPU_type ~__context ~self ~vgpu_type =
+        get_gpu_group_of_pgpu pgpu
+        |> get_gpu_types_of_gpu_group
       in
-      if supported then pgpu
-      else get_first_pgpu_of_implementation vgpu implementation rest
-  in
-  let test_nvidia_compatibility vgpu pgpu = 
-    let pgpu_metadata =
-      let key = Xapi_gpumon.Nvidia.key in
-      try
-        get_compatibility_metadata pgpu
-        |> List.assoc key
-      with Not_found -> 
-        debug "Key %s is missing from the compatibility_metadata for pgpu %s on the host %s." key (Ref.string_of pgpu) (Ref.string_of host);
-        raise Api_errors.(Server_error (nvidia_tools_error, [Ref.string_of host]))
-    in
-    Xapi_gpumon.Nvidia.assert_pgpu_is_compatible_with_vm ~__context ~vm ~vgpu ~dest_host:host ~encoded_pgpu_metadata:pgpu_metadata
+      if Listext.List.set_equiv pgpu_types types
+      then pgpu
+      else get_first_suitable_pgpu pgpu_types vgpu rest
   in
   let test_compatibility vgpu pgpus =
-    match vgpu_impl vgpu with
-      | `passthrough | `gvt_g | `mxgpu -> ()
-      | `nvidia as impl -> 
-        let pgpu = get_first_pgpu_of_implementation vgpu impl pgpus
-        in test_nvidia_compatibility vgpu pgpu
+    match vgpu_impl ~__context vgpu with
+    | `passthrough | `gvt_g | `mxgpu -> ()
+    | `nvidia ->
+      let pgpu_types =
+        Db.VGPU.get_GPU_group ~__context ~self:vgpu
+        |> fun self -> Db.GPU_group.get_GPU_types ~__context ~self
+      in
+      let pgpu = get_first_suitable_pgpu pgpu_types vgpu pgpus
+      in assert_destination_pgpu_is_compatible_with_vm ~__context ~vm ~vgpu ~pgpu ~host ?remote ()
   in
-  match vgpu_map with
-  | [] ->
-    let vgpus = Db.VM.get_VGPUs ~__context ~self:vm in
-    let pgpus = get_pgpus_of_host host in
-    List.iter (fun vgpu -> test_compatibility vgpu pgpus) vgpus
-  | vgpus ->
-    List.iter (fun (vgpu, gpu_group) ->
-      let pgpus = get_pgpus_of_gpu_group gpu_group in
+  let vgpus = Db.VM.get_VGPUs ~__context ~self:vm in
+  let _mapped, unmapped = List.partition (fun vgpu -> List.mem_assoc vgpu vgpu_map) vgpus in
+
+  (* Check that vgpus in the vgpu_map are correctly associable *)
+  List.iter (fun (vgpu, gpu_group) ->
+      match vgpu_impl ~__context vgpu with
+      | `passthrough | `gvt_g | `mxgpu -> ()
+      | `nvidia -> begin
+          let pgpus = get_pgpus_of_gpu_group gpu_group in
+          match pgpus with
+          | [] ->
+            debug "No compatible pGPU found using the provided vgpu_map, needed by VM %s" (Ref.string_of vm);
+            raise Api_errors.(Server_error (vm_requires_gpu, [Ref.string_of vm; Ref.string_of gpu_group]))
+          | pgpu :: _ ->
+            (* We can run the compatibility test with any of the PGPUs *)
+            assert_destination_pgpu_is_compatible_with_vm ~__context ~vm ~vgpu ~pgpu ~host ?remote ()
+        end
+    ) vgpu_map;
+
+  (* Check that there is a potential pgpu candidate for each of the other vgpus *)
+  List.iter (fun vgpu -> 
+      let pgpus = get_pgpus_of_host host in
```

Comment:
I think we don't need to obtain the `pGPUs` for every `vgpu` that we test - this binding can be moved out of `iter`.